server/src/main/java/org/elasticsearch/index/mapper/GeoShapeFieldMapper.java

```diff
@@ -517,20 +582,97 @@ public void parse(ParseContext context) throws IOException {
+            parseForPointsIndexing(context);
+        } else {
+            parseForInvertedIndexing(context);
+        }
+    }
+
     private void indexShape(ParseContext context, Shape shape) {
-        List<IndexableField> fields = new ArrayList<>(Arrays.asList(fieldType().defaultStrategy().createIndexableFields(shape)));
+        List<IndexableField> fields = new ArrayList<>(Arrays.asList(fieldType().defaultPrefixTreeStrategy().createIndexableFields(shape)));
         createFieldNamesField(context, fields);
         for (IndexableField field : fields) {
             context.doc().add(field);
         }
     }
 
+    private void indexShape(ParseContext context, Object luceneShape) {
+        if (luceneShape instanceof GeoPoint) {
+            GeoPoint pt = (GeoPoint)luceneShape;
+            indexFields(context, LatLonShape.createIndexableFields(name(), pt.lat(), pt.lon()));
+        } else if (luceneShape instanceof Line) {
+            indexFields(context, LatLonShape.createIndexableFields(name(), (Line)luceneShape));
+        } else if (luceneShape instanceof Polygon) {
+            indexFields(context, LatLonShape.createIndexableFields(name(), (Polygon) luceneShape));
+        } else if (luceneShape instanceof double[][]) {
+            double[][] pts = (double[][])luceneShape;
+            for (int i = 0; i < pts.length; ++i) {
+                indexFields(context, LatLonShape.createIndexableFields(name(), pts[i][1], pts[i][0]));
+            }
+        } else if (luceneShape instanceof Line[]) {
+            Line[] lines = (Line[]) luceneShape;
+            for (int i = 0; i < lines.length; ++i) {
+                indexFields(context, LatLonShape.createIndexableFields(name(), lines[i]));
+            }
+        } else if (luceneShape instanceof Polygon[]) {
+            Polygon[] polys = (Polygon[]) luceneShape;
+            for (int i = 0; i < polys.length; ++i) {
+                indexFields(context, LatLonShape.createIndexableFields(name(), polys[i]));
+            }
+        } else if (luceneShape instanceof Rectangle) {
+            // index rectangle as a polygon
+            Rectangle r = (Rectangle) luceneShape;
+            Polygon p = new Polygon(new double[]{r.minLat, r.minLat, r.maxLat, r.maxLat, r.minLat},
+                new double[]{r.minLon, r.maxLon, r.maxLon, r.minLon, r.minLon});
+            indexFields(context, LatLonShape.createIndexableFields(name(), p));
+        } else if (luceneShape instanceof Object[]) {
+            // recurse to index geometry collection
+            for (Object o : (Object[])luceneShape) {
+                indexShape(context, o);
+            }
+        } else {
+            throw new IllegalArgumentException("invalid shape type found [" + luceneShape.getClass().getName() + "] while indexing shape");
+        }
+    }
+
+    private void indexFields(ParseContext context, Field[] fields) {
+        ArrayList<IndexableField> flist = new ArrayList<>(Arrays.asList(fields));
```

Comment:
why are you making a copy instead of just `List<IndexableField> flist = Arrays.asList(fields)`?

Reply:
Because `createFieldNamesField` adds elements to the list, and Lists returned by `Arrays.asList` cannot be structurally modified.

Reply:
thanks I had missed this